tests/templates/cmake.toml

```diff
@@ -0,0 +1,18 @@
+# A project using templates.
+
+[project]
+name = "templates"
+description = "Template example"
+
+[template.app]
+type = "executable"
+sources = ["src/templates.cpp"]
+compile-definitions = ["IS_APP=true"]
+
+[target.app-a]
+template = "app"
```

Comment:
Perhaps it would be better to separate the parsing from the generation and instead change `target.type` to allow specifying a template?

The target parsing could be reused and the merge of the target fields could happen at generation time instead.

Reply:
Are you thinking something more along the lines of this?

```toml
# A project using templates.

[project]
name = "templates"
description = "Template example"

[target.app]
type = "template"
sources = ["src/templates.cpp"]
compile-definitions = ["IS_APP=true"]

[target.app-a]
template = "app"
type = "executable"
compile-definitions = ["APP_A"]

[target.app-b]
template = "app"
type = "executable"
compile-definitions = ["APP_B"]
```

As for reusing target parsing, this PR is already doing that, but I can see how doing something like the above could probably keep it simpler. I'd be fine implementing it however you wanted. Just let me know if the above guess is what you had in mind.

Reply:
> Perhaps it would be better to separate the parsing from the generation and instead change target.type to allow specifying a template?

If you could provide a little more guidance on (or an example of) what changes you're looking for I would be happy to work on them. 

> The target parsing could be reused

The way templates are currently implemented by this PR already has it reusing the same code for both template/target.

> the merge of the target fields could happen at generation time instead.

After looking at the generator code I'm not convinced that merging template with target at generation time would result in a simpler implementation of this feature.  Is there a compelling reason to do it here rather than at parse time? It seems to me like it would just complicate the generator, whereas doing the merge at parse time doesn't really change the parser at all.

Reply:
Been busy recently, but I've been thinking about this a bit and I'm not really sure what the best way to do it is yet.

Perhaps:

```toml
[template.app]
type = "executable"
sources = ["src/templates.cpp"]

[target.foo]
type = "app" # directly use the template name here (existing names should error)
sources = ["src/foo.cpp"]
```

The thing that bothers me about the current implementation is the `.optional_append` and I still think it's bad to mix parsing and generation. The parsers job should be to deserialize the toml file, preferably in a way that allows you to also serialize it back. When you start mixing the responsibilities you might save some lines, but what if you wanted to build a UI for `cmake.toml`?

With regards to the `GAME_NAME` you can use the `CMKR_TARGET` variable to a similar purpose.

In general I'm also still not 100% convinced about the utility of doing templates this way, maybe it would be enough to create a new target type `template` and inherit all the properties:

```toml
[target.app]
type = "template"
sources = ["src/templates.cpp"]

[target.foo]
type = "executable"
link-libraries = ["template"]
```

This wouldn't change the parsing code at all, just add a few additional lines to the generation. The downside is that `link-libraries` becomes even more overloaded and it might be more confusing than an explicit `template.`

Also a minor thing I think you made a mistake rebasing, because you duplicated my commits in your branch. Running `git rebase upstream/master` should get rid of these.

Reply:
Another thing that I’ve been considering is that a “common” CMake idiom seems to be to include a helper script that provides a function to create a target suitable for a plugin/whatever:

add_pintool(targetName ${SOURCES})

I think the WDK library does that, so does pybind11 and DynamoRIO. Ideally a template/custom target type should encompass that idiom.

Reply:
Thanks for the reply. Fair points on all of my questions, and thanks for the examples.